metadata_service/proxy/atlas_proxy.py

```diff
@@ -325,24 +333,30 @@ def get_column_description(self, *,
             column_name=column_name)
         return column_detail[self.ATTRS_KEY].get('description')
 
-    def get_popular_tables(self, *, num_entries: int) -> List[PopularTable]:
-        """
-        :param num_entries: Number of popular tables to fetch
-        :return: A List of popular tables instances
-        """
-        popular_tables = list()
+    @_CACHE.cache('_get_metadata_collection', _GET_POPULAR_TABLE_CACHE_EXPIRY_SEC)
+    def _get_metadata_collection(self, dsl_param: dict) -> List:
         try:
             # Fetch the metadata entities based on popularity score
-            query_metadata_ids = {'query': f'FROM Table SELECT metadata.__guid '
-                                           f'ORDERBY popularityScore desc '
-                                           f'LIMIT {num_entries}'}
-            metadata_ids = self._get_flat_values_from_dsl(dsl_param=query_metadata_ids)
+            metadata_ids = self._get_flat_values_from_dsl(dsl_param=dsl_param)
             metadata_collection = self._driver.entity_bulk(guid=metadata_ids)
+            return metadata_collection
         except KeyError as ex:
             LOGGER.exception(f'DSL Search query failed: {ex}')
             raise BadRequest('Unable to fetch popular tables. '
                              'Please check your configurations.')
 
+    def get_popular_tables(self, *, num_entries: int) -> List[PopularTable]:
+        """
+        :param num_entries: Number of popular tables to fetch
+        :return: A List of popular tables instances
+        """
+        popular_tables = list()
+        query_metadata_ids = {'query': f'FROM Table SELECT metadata.__guid '
+                                       f'ORDERBY popularityScore desc '
+                                       f'LIMIT {num_entries}'}
+
+        metadata_collection = self._get_metadata_collection(query_metadata_ids)
+
```

Comment:
Thanks @morpheyesh, this is really nice. 
I'd suggest to move the main getting entities logic to the cache function.
Like instead of _get_metadata_collection, your function should be _get_metadata_entities.
because essentially there would always be one item in the collection.

because this function `_collection.entities_with_relationships(attributes=["parentEntity"])` also need to make some queries to Atlas, which we don’t really want everytime.